### PR TITLE
Dns improvements

### PR DESCRIPTION
**Pull request by grobie**
See commit descriptions for more information.

### CODE REVIEW

**Comment by grobie** [comment]
Rebased (some changes got cherry-picked already into master).

Benefits:
* don't slow down DNS lookups for `chef` and other questions relying on search domains
* reduce (>50%) amount of requests glimpse needs to handle
* reduce (>>50%) number of invalid/unknown DNS requests in order to spot real errors faster

File: agent/dns.go [hunk 1]
```diff
@@ -146,6 +161,38 @@ func newRR(q dns.Question, i instance) dns.RR {
...
+		Retry:   600,
+		Expire:  86400,
+		Minttl:  defaultTTL,
+	}
+}
+
+func invalidDomain(q string) bool {
+	if q == "" {
+		return false
+	}
```
Comment (peterbourgon) [inline agent/dns.go]: Hmm, I'm a little confused by the boolean logic here. What if it were func validDomain(string) bool?

**Comment by peterbourgon** [comment]
> When search domains are used on client systems, the glimpse server will receive a lot of unknown questions. Responding with NOTIMPL will cause some recursors, e.g. unbound, to try again.

And what effect does NXDOMAIN have?

**Comment by peterbourgon** [comment]
Those few things aside, I'm :dango: to merge. But as I mentioned before, I'm still missing the motivation for this PR. 

Did you observe specific failure cases in the wild? If so, can you lay out what you saw? 

Or, are these changes best understood as bug fixes, to correctly implement the spec? If so, can we do anything to improve test coverage more systematically?

Or, are these just some inefficiencies that you noticed, and this PR is you scratching the itch? If so, can you provide a mini, NABC-style summary, so I can better put these individual changes into a broader context?

**Comment by xla** [comment]
:+1: :shipit: :octocat: :dango: 

Could you clarify how the day long timeout caching for authoritative domains would affect the introduction of new zones. Maybe this is based on wrong assumptions, but it is it correct that this could delay the discovery of a new zone, if it was requested before but didn't return a result?

**Comment by grobie** [comment]
@xla The `defaultInvalidTTL` only applies to domain formats glimpse-agent can not handle, e.g. anything not matching `<domain>.|<zone>.<domain>.|<service>.<job>.<env>.<product>.<zone>.<domain>.`. A request for `chef.<domain>.` is not valid as `chef` does not conform to the two character zone requirement. A request for `xx.<domain>.`, where `xx` is a not yet defined, is valid and will only be cached for `defaultTTL`.

Furthermore, at SoundCloud we have full control over our caching servers (unbound) and can flush their caches at any time.

**Comment by xla** [comment]
@grobie Thanks for the clarification. :dango:

File: agent/dns.go (hunk 1 above)
Comment (grobie) [inline agent/dns.go]: Done.

**Comment by grobie** [comment]
> And what effect does NXDOMAIN have?

Instead of retrying another 9 times and eventually returning SERVFAIL to the client, unbound will immediately return the NXDOMAIN response. So glimpse-agent will only see 1 request instead of 10 and clients will receive the response faster.

> Did you observe specific failure cases in the wild? If so, can you lay out what you saw?

Unexpected retries by unbound upon NOTIMPL responses as described above. It seems unbound treats any NOTIMPL response as an error and retries. Unbound will respond with SERVFAIL in this scenario, which can easily lead to wrong conclusions.

> Or, are these changes best understood as bug fixes, to correctly implement the spec?

The main motivation was to reduce the number of "errors" in both log and metrics. The signal/noise ratio (expected requests / unexpected/invalid requests) was improved from 0.17 to 1.64 in load tests.

**Comment by peterbourgon** [comment]
@grobie thanks for the context!

**Pull request merged**